book/src/trafficsim/parking.md

```diff
@@ -0,0 +1,14 @@
+# Parking
+
+TODO: Fill out the types of parking available, public/private, blackholes, how
```

Comment:
I know it's marked as "TODO", but can you give me a quick description of what blackhole means? It's used quite a lot in this code.

Reply:
Check out the "blackhole" layer. It's a lane that isn't reachable from all others, or that can't reach some other lanes. We use a [strongly-connected components](https://github.com/dabreegster/abstreet/blob/6e10f40801f7c720c4a44bc63cf56bedf6ce9744/map_model/src/connectivity.rs#L9) algorithm to find these. They tend to be around map borders. Imagine you're trying to park at a spot on a road leading off-map and there's no way to turn around. When someone later tries to drive the car from that spot somewhere else on the map, there isn't any path for them. So when looking for parking, we have to start before we reach that last lane. Similar for biking -- we should maybe hop onto the sidewalk a little early, because there's not a way to bike all the way.

Retrospective on this decision: it's caused moderate amount of headache in other layers, like here. Some of the cases would go away if a car could enter/exit a driveway from the off-side lane (aka, turn left into or out of a driveway). But implementing that correctly (blocking opposing lanes of traffic) is hard, because this sort of thing happens at intersections, and we don't have an intersection at every single driveway.

Reply:
That makes sense, thanks for clarifying.

So a black-holed building is a building accessible only via a blackholed lane?

Reply:
> So a black-holed building is a building accessible only via a blackholed lane?

Yes, in the `driving_blackhole` sense (not `biking_blackhole`)